### ISSUE TITLE
Backoff and retry instead of failing on throttling from azure

### ISSUE THREAD

**Issue by rjernst** [elastic/elasticsearch#22728]
Part of the motivation for #22709 was the azure repository plugin's snapshot failure when a throttling event occurs.  If we receive a throttling response, we should backoff and retry (and this may apply to our cloud code for other plugins as well).

**Comment by dadoonet** [comment]
I wonder if this should be done in core instead so no need to implement that for each plugin?

**Comment by rjernst** [comment]
I think it is only needed for cloud services. If there is an error from the filesystem, we should fail. Cloud systems are designed with this type of throttling in mind, but filesystems are not.

**Comment by dadoonet** [comment]
> Cloud systems are designed with this type of throttling in mind, but filesystems are not.

Yes you're absolutely right and I actually checked our code and I can see that Google Storage and AWS S3 sdks already provide this out of the box.
I did not check Azure SDK yet but if not provided in the SDK itself then we can implement our own version.

So quoting myself:

> I wonder if this should be done in core instead so no need to implement that for each plugin?

No. :)

**Comment by JeffreyZZ** [comment]
The current exception handling and its retry logical in Azure plugin is already able to handle the exceptions thrown when writing the data into Azure Storage account, it includes the throttling exception. What’s more, we also introduced ExponentialRetry to enhance the retry logic to make sure that it has enough retries when the exceptions happened in my PR#22709. But based on our investigation, this doesn’t help on the throttling scenarios, because when the snapshot data size is too big, such as 9TB in our case, it’s beyond the capability of exception retries because throttling lasts over a long period of time. Since the root cause is the bandwidth limit of a single storage account, I think that enabling the sharding of snapshot across multiple storage accounts is the right way to go for it. 

As we have consulted Azure Storage production team, they don't support accounts of higher IOPS, the only alternative solution is to shard the load across multiple storage accounts. I just added background investigation we did in PR#22709.

client.getDefaultRequestOptions().setRetryPolicyFactory(new RetryExponentialRetry(1000 * 30, 7));
